Material_Part4_Pitfalls/closure.py

```diff
@@ -1,12 +1,54 @@
 def make_multiplier_of(n):
     mylist = []
     for i in n:
+
         def multiplier(j):
             return i * j
+
         mylist.append(multiplier)
     return mylist
 
-times_x = make_multiplier_of(list(range(5)))
+
+def make_multiplier_of_A(n):
+    from functools import partial
+    mylist = []
+    for i in n:
+
+        def multiplier(factor1, factor2):
+            return factor1 * factor2
+
+        mylist.append(partial(multiplier, i))
+    return mylist
+
+
+# Question: Does partial do something like below?
+def make_multiplier_of_B(n):
+    mylist = []
+    for i in n:
+        def multiplier(factor1):
+            def multiply(factor2):
+                return factor2 * factor1
+            return multiply
+
+        mylist.append(multiplier(i))
+    return mylist
+
+
+def make_multiplier_of_C(n):
+    for i in n:
+
+        def multiplier(j):
+            return i * j
+
+        yield multiplier
+# Question: Are there other solutions?
```

Comment:
In your case you are returning n  multipliers again which would lead to the "pitfall"- behaviour. Instead if you use
```
def make_multiplier_of_C(n):
        def multiplier(j):
            return n * j
        yield multiplier
```
and then specify your n when generating the object:
```
multiply_3 = make_multiplier_of_C(3)
multiply3(8)